vendor/github.com/onsi/ginkgo/v2/internal/progress_reporter_manager.go

```diff
@@ -0,0 +1,79 @@
+package internal
+
+import (
+	"context"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/onsi/ginkgo/v2/types"
+)
+
+type ProgressReporterManager struct {
+	lock              *sync.Mutex
+	progressReporters map[int]func() string
+	prCounter         int
+}
+
+func NewProgressReporterManager() *ProgressReporterManager {
+	return &ProgressReporterManager{
+		progressReporters: map[int]func() string{},
+		lock:              &sync.Mutex{},
+	}
+}
+
+func (prm *ProgressReporterManager) AttachProgressReporter(reporter func() string) func() {
+	prm.lock.Lock()
+	defer prm.lock.Unlock()
+	prm.prCounter += 1
+	prCounter := prm.prCounter
+	prm.progressReporters[prCounter] = reporter
+
+	return func() {
+		prm.lock.Lock()
+		defer prm.lock.Unlock()
+		delete(prm.progressReporters, prCounter)
+	}
+}
+
+func (prm *ProgressReporterManager) QueryProgressReporters(ctx context.Context, failer *Failer) []string {
+	prm.lock.Lock()
+	keys := []int{}
+	for key := range prm.progressReporters {
+		keys = append(keys, key)
+	}
+	sort.Ints(keys)
+	reporters := []func() string{}
+	for _, key := range keys {
+		reporters = append(reporters, prm.progressReporters[key])
+	}
+	prm.lock.Unlock()
+
+	if len(reporters) == 0 {
+		return nil
+	}
+	out := []string{}
+	for _, reporter := range reporters {
+		reportC := make(chan string, 1)
+		go func() {
+			defer func() {
+				e := recover()
+				if e != nil {
+					failer.Panic(types.NewCodeLocationWithStackTrace(1), e)
+					reportC <- "failed to query attached progress reporter"
+				}
+			}()
+			reportC <- reporter()
+		}()
+		var report string
+		select {
+		case report = <-reportC:
+		case <-ctx.Done():
+			return out
+		}
+		if strings.TrimSpace(report) != "" {
+			out = append(out, report)
+		}
+	}
+	return out
+}
```

Comment:
The code looks generally well-written and there seem to be no obvious bugs. Some suggestions to improve the code might include:

1. Use a WaitGroup in the AttachProgressReporter function instead of a channel to make sure all goroutines have finished before returning.
2. Consider using a sync.RWMutex instead of a sync.Mutex to allow for concurrent reads when querying progress reporters.
3. Instead of using `reportC := make(chan string, 1)`, consider using an unbuffered channel and wrapping the select statement in a loop to ensure that each reporter is queried even if one blocks indefinitely.
4. Consider exposing the `progressReporters` map directly since it is already protected by the `lock`. This would allow external callers to iterate over the attached reporters without needing to create a separate function.